### ISSUE TITLE
Object oriented code?

### ISSUE THREAD

**Issue by clarkfitzg** [matloff/partools#11]
Base R methods such as `aggregate, split, mean, [` have corresponding functions `distribagg, distribsplit, distribmeans, distribgetrows` in partools.

It might be nice to have a distributed object, like a distributed data frame, and write the partools versions as methods.

Thoughts?

**Comment by matloff** [comment]
Yes, a subclass of 'data.frame', interesting.  What is your assessment
of the cost/benefit ratio (cost being development time)?

Norm

**Comment by clarkfitzg** [comment]
### Benefits

- Initially more familiar to users, which will help with adoption
- Well defined examples in base R to follow make it natural to extend this approach to other methods
- May simplify some partools code and make it more flexible if we can pass arguments directly through to methods on worker nodes using mechanism such as `...`

### Costs

- Likely a few weeks of initial coding
- Might end up creating and maintaining two versions of code which do the same thing

I've been thinking about this as I read Chambers' Extending R book (very good by the way).

I am just raising the issue to discuss and see if this is a possibility, or something that was considered and rejected.

**Comment by matloff** [comment]
I considered it briefly, but decided that there were too many technical
problems.  For example, compare the arguments of aggregate to
distribagg().

Norm

**Comment by clarkfitzg** [comment]
> For example, compare the arguments of aggregate to distribagg().

Could you explain further? Here's the relevant signatures:

```
## S3 method for class 'data.frame'
aggregate(x, by, FUN, ..., simplify = TRUE, drop = TRUE)

## S3 method for class 'formula'
aggregate(formula, data, FUN, ...,
           subset, na.action = na.omit)

## Partools:
distribagg(cls,ynames,xnames,dataname,FUN,FUNdim=1,FUN1=FUN)
```

**Comment by matloff** [comment]
Well, there you are.  The partools function uses strings, e.g. ynames,
and it requires two functions rather than one.  On top of that, some
features of aggregate() are not implemented and would be difficult to
add.

It would be useful if you were to add some distrib*() function yourself,
to see the problems.

Norm

**Comment by clarkfitzg** [comment]
> It would be useful if you were to add some distrib*() function yourself, to see the problems.

I just had a go at writing a function to add a new column to a dataframe which is already distributed, say `az`. The call was looking something like this:
```
distribsetcol(cls, dataname = az, colname = LETTER, FUN = function(x) toupper(x$letter))
```
where `FUN` is a function expecting the chunk of `az` in that node. Then I realized that it's cleaner and easier to instead use the stuff already in parallel: 
```
    clusterEvalQ(cls, {
        az$LETTER <- toupper(az$letter)
        NULL  # Necessary to avoid returning LETTER
    })
```
Maybe I'll add something like this into the docs.

**Comment by matloff** [comment]
Yes, a lot of useful things can be done directly using clusterEvalQ().

This should be emphasized in the vignette.  It would be nice if you put
your example into the vignette.

Norm

**Comment by clarkfitzg** [comment]
Example is in the vignette now.

**Comment by matloff** [comment]
Thanks.

Norm
